README


This archive contains the distribution MooseX-Role-Pluggable,
version 0.04:

  add plugins to your Moose classes

This software is copyright (c) 2013 by John SJ Anderson.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.




Changes
Revision history for MooseX-Role-Pluggable

0.04      2013-05-07 15:07:22 America/Los_Angeles

        Remove accidental MyApp::Plugin dep (RT#85111)

0.03      2013-04-07 13:07:08 America/Los_Angeles

        Make plugin_list ordered (Jody Clements)
        Convert to Dzil and generally clean up (John SJ Anderson)

0.2
0.1

        Undocumented. Bad maintainer, no cookie. 


Makefile.PL

use strict;
use warnings;

use 5.010;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "add plugins to your Moose classes",
  "AUTHOR" => "John SJ Anderson <genehack\@genehack.org>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "MooseX-Role-Pluggable",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "MooseX::Role::Pluggable",
  "PREREQ_PM" => {
    "Class::MOP" => 0,
    "Moose" => "2.000",
    "Moose::Role" => 0,
    "Moose::Util::TypeConstraints" => 0,
    "Tie::IxHash" => 0
  },
  "TEST_REQUIRES" => {
    "File::Find" => 0,
    "File::Temp" => 0,
    "FindBin" => 0,
    "Moose" => 0,
    "Test::More" => 0,
    "namespace::autoclean" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "VERSION" => "0.04",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  my $tr = delete $WriteMakefileArgs{TEST_REQUIRES};
  my $br = $WriteMakefileArgs{BUILD_REQUIRES};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/MooseX/Role/Pluggable/Plugin.pm
package MooseX::Role::Pluggable::Plugin;
{
  $MooseX::Role::Pluggable::Plugin::VERSION = '0.04';
}
BEGIN {
  $MooseX::Role::Pluggable::Plugin::AUTHORITY = 'cpan:GENEHACK';
}
# ABSTRACT: Role for plugins to consume
use Moose::Role;
use Moose::Util::TypeConstraints;

has name => (
  isa => 'Str' ,
  is  => 'ro' ,
  required => 1 ,
);

subtype 'MooseXRolePluggable'
  => as 'Object'
  => where { $_->does( 'MooseX::Role::Pluggable' ) }
  => message { 'Parent does not consume MooseX::Role::Pluggable!' };

has parent => (
  isa      => 'MooseXRolePluggable' ,
  is       => 'ro' ,
  required => 1 ,
);

no Moose::Role;
1;

__END__

=pod

=head1 NAME

MooseX::Role::Pluggable::Plugin - Role for plugins to consume

=head1 VERSION

version 0.04

=head1 SYNOPSIS

    package MyMoose::Plugin::Antlers;
    use Moose;
    with 'MooseX::Role::Pluggable::Plugin';

    sub gore {
      my( $self ) = shift;

      # yadda yadda yadda
    }

    # see the documentation for MooseX::Role::Pluggable for info on how to get
    # your Moose class to use this plugin...

=head1 DESCRIPTION

This is a role that must be consumed by any plugin that's going to be used
with the 'MooseX::Role::Pluggable' role. It serves to make sure that required
attributes are in place and contain the expected sorts of info.

Classes that want to utilize plugins should consume the
'MooseX::Role::Pluggable' role; see the documentation for that module for more
information.

=head1 NAME

MooseX::Role::Pluggable::Plugin - add plugins to your Moose classes

=head1 AUTHOR

John SJ Anderson, C<genehack@genehack.org>

=head1 SEE ALSO

L<MooseX::Role::Pluggable>

=head1 COPYRIGHT AND LICENSE

Copyright 2010, John SJ Anderson

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=head1 AUTHOR

John SJ Anderson <genehack@genehack.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by John SJ Anderson.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/MooseX/Role/Pluggable.pm
package MooseX::Role::Pluggable;
{
  $MooseX::Role::Pluggable::VERSION = '0.04';
}
BEGIN {
  $MooseX::Role::Pluggable::AUTHORITY = 'cpan:GENEHACK';
}
# ABSTRACT: add plugins to your Moose classes
use Class::MOP;
use Moose::Role;
use Moose::Util::TypeConstraints;
use Tie::IxHash;
use 5.010;

has plugins => (
  isa => 'ArrayRef[Str]',
  is  => 'rw' ,
);

subtype 'MooseXRolePluggablePlugin'
  => as 'Object'
  => where { $_->does( 'MooseX::Role::Pluggable::Plugin' ) }
  => message { 'Plugin did not consume the required role!' };

has plugin_hash => (
  isa        => 'Maybe[HashRef[MooseXRolePluggablePlugin]]' ,
  is         => 'ro' ,
  init_arg   => undef ,
  lazy_build => 1 ,
);

sub _build_plugin_hash {
  my $self = shift;

  return $self->plugin_list ? { map { $_->name => $_ } @{ $self->plugin_list } } : undef;
}

has plugin_list => (
  isa        => 'Maybe[ArrayRef[MooseXRolePluggablePlugin]]' ,
  is         => 'ro' ,
  init_arg   => undef ,
  lazy_build => 1 ,
);

sub _build_plugin_list {
  my( $self ) = shift;

  return undef unless $self->plugins;

  my $plugin_list = [];

  my $plugin_name_map = $self->_map_plugins_to_libs();

  foreach my $plugin_name ( keys %$plugin_name_map ) {
    my $plugin_lib = $plugin_name_map->{$plugin_name};

    ### FIXME should have some Try::Tiny here, with a parameter to control
    ### what happens when a class doesn't load -- ignore, warn, die
    Class::MOP::load_class( $plugin_lib );

    my $plugin = $plugin_lib->new({
      name   => $plugin_name ,
      parent => $self ,
    });

    push @{ $plugin_list } , $plugin;
  }

  return $plugin_list;
};

sub plugin_run_method {
  my( $self , $method ) = @_;

  my $return = [];
  foreach my $plugin ( @{ $self->plugin_list }) {
    if ( $plugin->can( $method ) ) {
      push @$return , $plugin->$method();
    }
  }
  return $return;
}

sub _map_plugins_to_libs {
  my( $self ) = @_;
  my $class = ref $self;

  tie my %map, "Tie::IxHash";
  foreach ( @{ $self->plugins } ) {
    $map{$_} = ( s/^\+// ) ? $_ : "${class}::Plugin::$_";
  }
  return \%map;
}

no Moose::Role;
1;

__END__

=pod

=head1 NAME

MooseX::Role::Pluggable - add plugins to your Moose classes

=head1 VERSION

version 0.04

=head1 SYNOPSIS

    package MyMoose;
    use Moose;
    with 'MooseX::Role::Pluggable';

    my $moose = MyMoose->new({
      plugins => [ 'Antlers' , 'Tail' , '+After::Market::GroundEffectsPackage' ] ,
      # other args here
    });

    foreach my $plugin ( @{ $moose->plugin_list } ) {
      if ( $plugin->can( 'some_method' )) {
        $plugin->some_method();
      }
    }

    # call a method in a particular plugin directly
    # (plugin_hash() returns a hash ref of 'plugin_name => plugin')
    $moose->plugin_hash->{Antlers}->gore( $other_moose );

    # plugins are indexed by the name that was used in the original 'plugins' list
    $moose->plugin_hash->{After::Market::GroundEffectsPackage}->light_up();

    # see the documentation for MooseX::Role::Pluggable::Plugin for info on
    # how to write plugins...

=head1 DESCRIPTION

This is a role that allows your class to consume an arbitrary set of plugin
modules and then access those plugins and use them to do stuff.

Plugins are loaded based on the list of plugin names in the 'plugins'
attribute. Names that start with a '+' are used as the full name to load;
names that don't start with a leading '+' are assumed to be in a 'Plugins'
namespace under your class name. (E.g., if your app is 'MyApp', plugins will
be loaded from 'MyApp::Plugin').

NOTE: Plugins are lazily loaded -- that is, no plugins will be loaded until
either the 'plugin_list' or 'plugin_hash' methods are called. If you want to
force plugins to load at object instantiation time, your best bet is to call
one of those method right after you call 'new()'.

Plugin classes should consume the 'MooseX::Role::Pluggable::Plugin' role; see
the documentation for that module for more information.

=head1 NAME

MooseX::Role::Pluggable - add plugins to your Moose classes

=head1 METHODS

=head2 plugin_hash

Returns a hashref with a mapping of 'plugin_name' to the actual plugin object.

=head2 plugin_list

Returns an arrayref of loaded plugin objects. The arrayref will have the
same order as the plugins array ref passed in during object creation.

=head2 plugin_run_method( $method_name )

Looks for a method named $method_name in each plugin in the plugin list. If a
method of given name is found, it will be run. (N.b., this is essentially the
C<foreach> loop from the L</SYNOPSIS>.)

=head1 AUTHOR

John SJ Anderson, C<genehack@genehack.org>

=head1 SEE ALSO

L<MooseX::Role::Pluggable::Plugin>, L<MooseX::Object::Pluggable>,
L<Object::Pluggable>

=head1 COPYRIGHT AND LICENSE

Copyright 2010, John SJ Anderson

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=head1 AUTHOR

John SJ Anderson <genehack@genehack.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by John SJ Anderson.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
